lib/util/index.ts

```diff
@@ -0,0 +1,270 @@
+import * as errors from 'balena-errors';
+import assign = require('lodash/assign');
+import cloneDeep = require('lodash/cloneDeep');
+import isArray = require('lodash/isArray');
+import isFunction = require('lodash/isFunction');
+import isNumber = require('lodash/isNumber');
+import isString = require('lodash/isString');
+import throttle = require('lodash/throttle');
+import * as memoizee from 'memoizee';
+import * as moment from 'moment';
+import BalenaSdk = require('../../typings/balena-sdk');
+import { getOsUpdateHelper as updateHelper } from './device-actions/os-update';
+import * as dt from './device-types';
+
+export const getOsUpdateHelper = updateHelper;
+export const deviceTypes = dt;
+
+export const notImplemented = () => {
+	throw new Error('The method is not implemented.');
+};
+
+export const onlyIf = (condition: boolean) => (fn: () => void) => {
+	if (condition) {
+		return fn;
+	} else {
+		return notImplemented;
+	}
+};
+
+export const now = throttle(() => moment(), 1000, { leading: true });
+
+export const dateToMoment = memoizee((date: Date) => moment(date), {
+	max: 1000,
+	primitive: true,
+});
+
+export const timeSince = (input: Date, suffix = true) => {
+	const date = dateToMoment(input);
+
+	// We do this to avoid out-of-sync times causing this to return
+	// e.g. 'in a few seconds'.
+	// if the date is in the future .min will make it at maximum, the time since now
+	// which results in 'a few seconds ago'.
+	const time = now();
+	return moment.min(time, date).from(time, !suffix);
+};
+
+export const isId = isNumber;
+
+export const LOCKED_STATUS_CODE = 423;
+
+// Use with: `findCallback(arguments)`.
+export const findCallback = (args: IArguments) => {
+	const lastArg = args[args.length - 1];
+	if (isFunction(lastArg)) {
+		return lastArg;
+	}
+
+	return;
+};
+
+export const unauthorizedError = {
+	code: 'BalenaRequestError',
+	statusCode: 401,
+};
+
+export const notFoundResponse = {
+	code: 'BalenaRequestError',
+	statusCode: 404,
+};
+
+export const noDeviceForKeyResponse = {
+	code: 'BalenaRequestError',
+	statusCode: 500,
+	body: 'No device found to associate with the api key',
+};
+
+export const noApplicationForKeyResponse = {
+	code: 'BalenaRequestError',
+	statusCode: 500,
+	body: 'No application found to associate with the api key',
+};
+
+export const isUniqueKeyViolationResponse = ({
+	code,
+	body,
+}: {
```

Comment:
Probably worth making an interface for these custom error objects and using it here